library/ssl_msg.c

```diff
@@ -4135,7 +4135,12 @@ static int ssl_prepare_record_content(mbedtls_ssl_context *ssl,
         if (rec->type == MBEDTLS_SSL_MSG_APPLICATION_DATA) {
             MBEDTLS_SSL_DEBUG_MSG(
                 3, ("EarlyData: Ignore application message before 2nd ClientHello"));
-            /* TODO: Add max_early_data_size check here, see issue 6347 */
+
+            ret = mbedtls_ssl_tls13_check_early_data_len(ssl, rec->data_len);
```

Comment:
Same here, we can't deprotect the records so we can't know the padding length, but it seems to me we could still subtract AEAD tag length + 1 for the content type byte. That way, we minimize the risk of terminating the handshake when the client actually respected the limit. (Also, it would allow exact tests in builds without padding.)

Reply:
Here the AEAD tag length is indeed not substracted. We are in the HRR case, we do not have the handshake traffic keys and thus the handshake code has not set an inbound transform. A ciphersuite has been selected though but this information is to me an handshake layer information and we are here in the record layer. I know that currently we have access to any information we need in any layer but I am not very comfortable to use it here anyway (with MPS that I think the kind of things we will not be able to do anymore). Thus I would rather be on the side of keeping this as it is.
It is true that even if a client respect max_early_data_size on its side the server may consider it has sent too much early data and abort the handshake but anyway something was not completely clean in the first place: use of a ticket and need for an HRR. What do you think?

Reply:
Ok, let's not make things overly complicated just for this.